Common/include/mpi_structure.hpp

```diff
@@ -180,101 +230,112 @@ class CAuxMPIWrapper : public CMPIWrapper{
-                    int tag, MPI_Comm comm, MPI_Request* request);
+  static void Irecv(void *buf, int count, Datatype datatype, int source,
+                    int tag, Comm comm, Request* request);
 
-  static void Wait(MPI_Request *request, MPI_Status *status);
+  static void Wait(Request *request, Status *status);
 
-  static void Waitall(int nrequests, MPI_Request *request, MPI_Status *status);
+  static void Waitall(int nrequests, Request *request, Status *status);
 
-  static void Waitany(int nrequests, MPI_Request *request,
-                      int *index, MPI_Status *status);
+  static void Waitany(int nrequests, Request *request,
+                      int *index, Status *status);
 
-  static void Send(void *buf, int count, MPI_Datatype datatype, int dest,
-                   int tag, MPI_Comm comm);
+  static void Send(void *buf, int count, Datatype datatype, int dest,
+                   int tag, Comm comm);
 
-  static void Recv(void *buf, int count, MPI_Datatype datatype, int dest,
-                   int tag, MPI_Comm comm, MPI_Status *status);
+  static void Recv(void *buf, int count, Datatype datatype, int dest,
+                   int tag, Comm comm, Status *status);
 
-  static void Bcast(void *buf, int count, MPI_Datatype datatype, int root,
-                    MPI_Comm comm);
+  static void Bcast(void *buf, int count, Datatype datatype, int root,
+                    Comm comm);
 
-  static void Bsend(void *buf, int count, MPI_Datatype datatype, int dest,
-                    int tag, MPI_Comm comm);
+  static void Bsend(void *buf, int count, Datatype datatype, int dest,
+                    int tag, Comm comm);
 
   static void Reduce(void *sendbuf, void *recvbuf, int count,
-                     MPI_Datatype datatype, MPI_Op op, int root, MPI_Comm comm);
+                     Datatype datatype, Op op, int root, Comm comm);
 
   static void Allreduce(void *sendbuf, void *recvbuf, int count,
-                        MPI_Datatype datatype, MPI_Op op, MPI_Comm comm);
+                        Datatype datatype, Op op, Comm comm);
 
-  static void Gather(void *sendbuf, int sendcnt,MPI_Datatype sendtype,
-                     void *recvbuf, int recvcnt, MPI_Datatype recvtype, int root, MPI_Comm comm);
+  static void Gather(void *sendbuf, int sendcnt, Datatype sendtype,
+                     void *recvbuf, int recvcnt, Datatype recvtype, int root, Comm comm);
 
-  static void Scatter(void *sendbuf, int sendcnt,MPI_Datatype sendtype,
-                      void *recvbuf, int recvcnt, MPI_Datatype recvtype, int root, MPI_Comm comm);
+  static void Scatter(void *sendbuf, int sendcnt, Datatype sendtype,
+                      void *recvbuf, int recvcnt, Datatype recvtype, int root, Comm comm);
 
-  static void Allgather(void *sendbuf, int sendcnt, MPI_Datatype sendtype,
-                        void *recvbuf, int recvcnt, MPI_Datatype recvtype, MPI_Comm comm);
+  static void Allgather(void *sendbuf, int sendcnt, Datatype sendtype,
```

Comment:
Yep, all of this collective communications should work now. They just have to be added to the wrapper.

Reply:
In my opinion MPI_Allgatherv is rather essential to have, because the current workaround can be very memory consuming, especially when the number of ranks increases. There are other MPI routines that are used in the DG branches that are (were) not supported, but these are not used with su2double. So it would be nice to have them supported for consistency, but it does not have the highest priority. These are
- MPI_Reduce_scatter
- MPI_Probe
- MPI_Get_count
- MPI_Alltoall
- MPI_Testall

And then there are still the calls to MPI_Buffer_attach and MPI_Buffer_detach, which I have never understood.

Reply:
Actually I can remove the MPI_Buffer_attach/detach. I was using it before for my naive implementation of the forward mode MPI wrapper.

I will add the routines your are referring to.